net/instaweb/http/redirect_following_url_async_fetcher.cc

```diff
@@ -0,0 +1,333 @@
+          response_headers()->SetCacheControlMaxAge(max_age_);
+        }
+      }
+      SharedAsyncFetch::HandleHeadersComplete();
+    } else {
+      received_redirect_status_code_ = true;
+    }
+  }
+
+  bool HandleWrite(const StringPiece& content,
+                           MessageHandler* handler) override {
+    if (!received_redirect_status_code_) {
+      return SharedAsyncFetch::HandleWrite(content, handler);
+    }
+    return true;
+  }
+
+  void HandleDone(bool success) override {
+    DCHECK(gurl_.IsWebValid() || !success);
+
+    if (!received_redirect_status_code_) {
+      SharedAsyncFetch::HandleDone(success);
+      delete this;
+      return;
+    }
+
+    GoogleString redirect_url;
+    GoogleString mapped_redirect_url;
+
+    if (success) {
+      success = TryExtractRedirectUrlFromResponseHeaders(redirect_url);
+    }
+    if (success) {
+      success = CheckRedirectAdministration(redirect_url);
+    }
+    if (success) {
+      success = TryMapRedirect(redirect_url, mapped_redirect_url);
+    }
+
+    // Wipe out the 3XX response. We'll either fail/404 or return 200/OK
+    response_headers()->Clear();
+
+    if (success) {
+      redirect_following_fetcher_->FollowRedirect(
+          mapped_redirect_url, message_handler_, base_fetch_,
+          urls_seen_.release(), max_age_);
+    } else {
+      response_headers()->set_status_code(HttpStatus::kNotFound);
+      SharedAsyncFetch::HandleDone(false);
+    }
+
+    delete this;
+  }
+
+ private:
+  void EmitRedirectWarning(const GoogleString& context_url,
+                           const GoogleString& redirect_url,
+                           const GoogleString message) {
+      const char* message_template = "Fetch redirect: [%s] -> [%s]: %s.";
+      message_handler_->Message(
+          kWarning, message_template, context_url.c_str(),
+          redirect_url.c_str(), message.c_str());
+  }
+
+  bool CheckRedirectAdministration(const GoogleString& redirect_url) {
+    int redirects_followed = urls_seen_->size();
+
+    if (redirects_followed > redirect_following_fetcher_->max_redirects()) {
+      EmitRedirectWarning(url_, redirect_url, "Max redirects exceeded");
+      return false;
+    }
+
+    std::pair<StringSet::iterator, bool> ret =
+        urls_seen_->insert(redirect_url);
+    if (!ret.second) {
+      EmitRedirectWarning(url_, redirect_url, "Cyclic redirect detected");
+      return false;
+    }
+    return true;
+  }
+
+  bool TryMapRedirect(const GoogleString& redirect_url,
+                           GoogleString& mapped_url) {
+
+    const RewriteOptions* options =
+        redirect_following_fetcher_->rewrite_options();
+    const DomainLawyer* domain_lawyer = options->domain_lawyer();
+
+    bool cacheable = response_headers()->IsProxyCacheable(
+        base_fetch_->request_headers()->GetProperties(),
+        ResponseHeaders::GetVaryOption(options->respect_vary()),
+        ResponseHeaders::kNoValidator);
+
+    if (!cacheable) {
```

Comment:
Hmm, this + max_age_ covers the major cases, but it may be worth checking what happens with 
the subtler stuff like must-revalidate (which we probably can't handle well, come to think of it)  and 
s-maxage...